File: 백준/Gold/10942. 팰린드롬？/팰린드롬？.py
import sys
input = sys.stdin.readline
n = int(input())
arr = [*map(int,input().split())]
dp = [[False] * n for _ in range(n)]

for i in range(n):
    for start in range(n):
        end = start + i
        if end>=n: #길이를 넘으면 중지
            break
        if i ==0: #길이가 1 일때는 무조건 팰린드롬
            dp[start][end] = True
            continue
        if i == 1: #길이가 2이면 시작과 끝이 같으면 팰린드롬
            if arr[start] == arr[end]:
                dp[start][end] = True
            continue
        if arr[start] == arr[end] and dp[start+1][end-1]: # 만약 시작과 끝이 같고 그 안이 팰린드롬이면 팰린드롬롬            dp[start][end] = True
            dp[start][end] = True
m = int(input())
for _ in range(m):
    a,b = map(int,input().split())
    if dp[a-1][b-1]:
        print(1)
    else:
        print(0)

File: 백준/Gold/10942. 팰린드롬？/팰린드롬？.rs
use std::process::exit;
use std::{io, vec};
use std::io::prelude::*;

fn read<T>(si: &mut T) -> String where T: Read {
    let mut s = String::new();
    si.read_to_string(&mut s).unwrap();
    s
}

fn next<T>(it: &mut std::str::SplitAsciiWhitespace) -> T where
    T: std::str::FromStr,
    <T as std::str::FromStr>::Err: std::fmt::Debug {
    it.next().unwrap().parse().unwrap()
}


fn main() {
    // 입력
    let mut si = io::BufReader::new(io::stdin().lock());
    let mut so = io::BufWriter::new(io::stdout().lock());
    let s = read(&mut si);
    let mut it = s.split_ascii_whitespace();
    let n = next::<usize>(&mut it);
    let arr = (0..n).map(|_|
    next::<usize>(&mut it)).collect::<Vec<usize>>();
    let mut dp = vec![vec![false; arr.len()]; arr.len()];

    for l in (0..n) {
        for start in (0..n-l) {
            let end = start + l;
            if start == end {
                dp[start][end] = true;
            }
            else if arr[start] == arr[end] {
                if start + 1 == end {
                    dp[start][end] = true;
                }
                else {
                    dp[start][end] = dp[start+1][end-1];
                }
            }
        }
    }
    for i in (0..next::<usize>(&mut it)) {
        let start = next::<usize>(&mut it) - 1;
        let end = next::<usize>(&mut it) - 1;
        if dp[start][end] {
            writeln!(so, "1").ok();
        }
        else {
            writeln!(so, "0").ok();
        }
    }
}